src/com/facebook/buck/cxx/CxxLibraryFactory.java

```diff
@@ -413,17 +413,23 @@ public BuildRule createBuildRule(
     }
 
     // Create rule to build the object files.
-    ImmutableMultimap<CxxSource.Type, Arg> compilerFlags =
-        ImmutableListMultimap.copyOf(
-            Multimaps.transformValues(
-                CxxFlags.getLanguageFlagsWithMacros(
-                    args.getCompilerFlags(),
-                    args.getPlatformCompilerFlags(),
-                    args.getLangCompilerFlags(),
-                    cxxPlatform),
-                f ->
-                    CxxDescriptionEnhancer.toStringWithMacrosArgs(
-                        buildTarget, cellRoots, ruleResolver, cxxPlatform, f)));
+    ImmutableListMultimap.Builder<CxxSource.Type, Arg> allCompilerFlagsBuilder =
+        ImmutableListMultimap.builder();
+    allCompilerFlagsBuilder.putAll(
+        Multimaps.transformValues(
+            CxxFlags.getLanguageFlagsWithMacros(
+                args.getCompilerFlags(),
+                args.getPlatformCompilerFlags(),
+                args.getLangCompilerFlags(),
+                cxxPlatform),
+            f ->
+                CxxDescriptionEnhancer.toStringWithMacrosArgs(
+                    buildTarget, cellRoots, ruleResolver, cxxPlatform, f)));
+    if (args.isModular()) {
+      allCompilerFlagsBuilder.putAll(
+          CxxFlags.toLanguageFlags(StringArg.from("-fmodule-name=" + args.getModuleName().get())));
```

Comment:
I'm not sure this is the correct place to put this. It seems like this might be better in the `CxxSourceRuleFactory`, but I couldn't find a simple place to inject it. This seemed the easiest place but I'm having a bit of trouble writing a reasonable test for it now, which would suggest that this isn't the best place.

The current suggestion I have now is passing along the modulename to `CxxSourceRuleFactory` if modular is set and it is present, to then add it to the flags through `private Iterable<Arg> getRuleCompileFlags(CxxSource.Type type)` there.

Reply:
@robbertvanginkel:
> I'm not sure this is the correct place to put this. It seems like this might be better in the CxxSourceRuleFactory, but I couldn't find a simple place to inject it.

There's at least one other place that might need this logic: `CxxDescriptionEnhancer::createBuildRulesForCxxBinary` which contains the following code (which is almost 1:1 duplicated):

```java
ImmutableListMultimap.copyOf(
  Multimaps.transformValues(
      CxxFlags.getLanguageFlagsWithMacros(
          preprocessorFlags,
          platformPreprocessorFlags,
          langPreprocessorFlags,
          cxxPlatform),
      f -> toStringWithMacrosArgs(target, cellRoots, resolver, cxxPlatform, f))),
```

Adding it as a `@Value.Parameter` to `AbstractCxxSourceRuleFactory` would make sure that we definitely add the module name parameter to all source rules rather than relying on hunting down all call sites.

Having said that, I'm not sure whether it's useful to pass `-fmodule-name` for rules like `apple_binary`. Do we know that we want to add the compiler flag everywhere? Another aspect that comes to mind is whether we want to have this flag added for (Objective-)C++ files where modules are not yet finalised.